<a href="https://colab.research.google.com/github/sjtae/data_science_project/blob/main/project01/Python_Project/FitBit_Data_Tracker_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# FitBit Fitness Tracker Data Analysis


In [44]:
# Libraries
from pathlib import Path
import sqlite3
import pandas as pd

## Data Preprocessing

### Create database from several csv files

In [5]:
Path('my_database.db').touch()

In [6]:
conn = sqlite3.connect('my_database.db')
c = conn.cursor()

In [7]:
file_names = [
    'dailyActivity_merged.csv',
    'dailyCalories_merged.csv',
    'dailyIntensities_merged.csv',
    'dailySteps_merged.csv',
    'heartrate_seconds_merged.csv',
    'hourlyCalories_merged.csv',
    'hourlyIntensities_merged.csv',
    'hourlySteps_merged.csv',
    'minuteCaloriesNarrow_merged.csv',
    'minuteCaloriesWide_merged.csv',
    'minuteIntensitiesNarrow_merged.csv',
    'minuteIntensitiesWide_merged.csv',
    'minuteMETsNarrow_merged.csv',
    'minuteSleep_merged.csv',
    'minuteStepsNarrow_merged.csv',
    'minuteStepsWide_merged.csv',
    'sleepDay_merged.csv',
    'weightLogInfo_merged.csv'
]


data = {}  # Create an empty dictionary to store the data frames

for file_name in file_names:
    # Extract the variable name from the file name (remove ".csv" and "daily" prefix)
    var_name = file_name.replace('.csv', '')

    # Read the CSV file and store it in the dictionary
    data[var_name] = pd.read_csv(file_name)

    # Write the data frame to the SQLite database
    data[var_name].to_sql(f'{var_name}', conn, if_exists='replace', index=False)

# Close the database connection
# conn.close()


In [8]:
cursor = conn.cursor()
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")

# Fetch all the table names
table_names = cursor.fetchall()

# Iterate through the table names and print them
for name in table_names:
    print(name[0])

dailyActivity_merged
dailyCalories_merged
dailyIntensities_merged
dailySteps_merged
heartrate_seconds_merged
hourlyCalories_merged
hourlyIntensities_merged
hourlySteps_merged
minuteCaloriesNarrow_merged
minuteCaloriesWide_merged
minuteIntensitiesNarrow_merged
minuteIntensitiesWide_merged
minuteMETsNarrow_merged
minuteSleep_merged
minuteStepsNarrow_merged
minuteStepsWide_merged
sleepDay_merged
weightLogInfo_merged


## Data Analysis

### Descriptive statistics for steps and distance each person

In [9]:
pd.read_sql('''SELECT
                Id,
                COUNT(*) AS record_count,
                SUM(TotalSteps) AS total_steps,
                AVG(TotalSteps) AS average_steps,
                MIN(TotalSteps) AS min_steps,
                MAX(TotalSteps) as max_steps,
                SUM(TotalDistance) AS total_distance,
                AVG(TotalDistance) AS average_distance,
                MIN(TotalDistance) AS min_distance,
                MAX(TotalDistance) as max_distance
               FROM dailyActivity_merged
               GROUP BY 1''', conn)

Id  record_count  total_steps  average_steps  min_steps  \
0   1503960366            31       375619   12116.741935          0   
1   1624580081            31       178061    5743.903226       1510   
2   1644430081            30       218489    7282.966667       1223   
3   1844505072            31        79982    2580.064516          0   
4   1927972279            31        28400     916.129032          0   
5   2022484408            31       352490   11370.645161       3292   
6   2026352035            31       172573    5566.870968        254   
7   2320127002            31       146223    4716.870968        772   
8   2347167796            18       171354    9519.666667         42   
9   2873212765            31       234229    7555.774194       2524   
10  3372868164            20       137233    6861.650000       3077   
11  3977333714            30       329537   10984.566667        746   
12  4020332650            31        70284    2267.225806          0   
13  4057192912             4        15352    3838.000000          0   
14  4319703577            31       225334    7268.838710         17   
15  4388161847            31       335232   10813.935484       3369   
16  4445114986            31       148693    4796.548387        768   
17  4558609924            31       238239    7685.129032       3428   
18  4702921684            31       265734    8572.064516          0   
19  5553957443            31       266990    8612.580645        655   
20  5577150313            30       249133    8304.433333          0   
21  6117666160            28       197308    7046.714286          0   
22  6290855005            29       163837    5649.551724          0   
23  6775888955            26        65512    2519.692308          0   
24  6962181067            31       303639    9794.806452       1551   
25  7007744171            26       294409   11323.423077          0   
26  7086361926            31       290525    9371.774194          0   
27  8053475328            31       457662   14763.290323       1170   
28  8253242879            19       123161    6482.157895          0   
29  8378563200            31       270249    8717.709677       2132   
30  8583815059            31       223154    7198.516129          0   
31  8792009665            29        53758    1853.724138          0   
32  8877689391            31       497241   16040.032258       4790   

    max_steps  total_distance  average_distance  min_distance  max_distance  
0       18134      242.099999          7.809677          0.00     12.210000  
1       36019      121.360001          3.914839          0.98     28.030001  
2       18213      158.860001          5.295333          0.89     13.240000  
3        8054       52.890000          1.706129          0.00      5.320000  
4        3790       19.670000          0.634516          0.00      2.620000  
5       18387      250.609998          8.084193          2.31     12.910000  
6       12357      107.100000          3.454839          0.16      7.710000  
7       10725       98.819999          3.187742          0.52      7.490000  
8       22244      114.400000          6.355556          0.03     15.080000  
9        9685      158.149999          5.101613          1.70      6.650000  
10       9715       94.140001          4.707000          2.10      6.630000  
11      16520      225.509998          7.517000          0.50     11.050000  
12      11728       50.410000          1.626129          0.00      8.430000  
13       5974       11.450000          2.862500          0.00      4.470000  
14      13658      151.659999          4.892258          0.01      9.490000  
15      22770      260.190003          8.393226          2.59     17.540001  
16       9105      100.620000          3.245806          0.52      6.110000  
17      13743      157.500000          5.080645          2.27      9.080000  
18      15126      215.610000          6.955161          0.00     12.270000  
19      17022      174.830001        

### Average length of a step for each person

In [10]:
pd.read_sql('''SELECT
                Id,
                (AVG(TotalDistance) / AVG(TotalSteps)) * 100000 as step_length_cm
               FROM dailyActivity_merged
               GROUP BY 1
               ''', conn)

Id  step_length_cm
0   1503960366       64.453608
1   1624580081       68.156419
2   1644430081       72.708466
3   1844505072       66.127379
4   1927972279       69.260563
5   2022484408       71.097052
6   2026352035       62.060693
7   2320127002       67.581707
8   2347167796       66.762375
9   2873212765       67.519393
10  3372868164       68.598661
11  3977333714       68.432376
12  4020332650       71.723294
13  4057192912       74.583117
14  4319703577       67.304534
15  4388161847       77.614906
16  4445114986       67.669628
17  4558609924       66.110083
18  4702921684       81.137528
19  5553957443       65.481854
20  5577150313       74.819474
21  6117666160       75.810409
22  6290855005       75.623944
23  6775888955       71.971546
24  6962181067       67.237740
25  7007744171       70.785879
26  7086361926       68.162809
27  8053475328       77.727667
28  8253242879       72.003313
29  8378563200       79.304642
30  8583815059       78.008908
31  8792009665       64.009077
32  8877689391       82.374542

### How much calories does a step burn for each person?

In [11]:
pd.read_sql('''SELECT
                da.Id,
                AVG(da.TotalSteps) / AVG(dc.Calories) AS calories_burn_for_a_step
               FROM dailyActivity_merged da
               JOIN dailyCalories_merged dc
               ON da.Id = dc.Id
               GROUP BY 1
               ''', conn)

Id  calories_burn_for_a_step
0   1503960366                  6.670674
1   1624580081                  3.872238
2   1644430081                  2.590605
3   1844505072                  1.639715
4   1927972279                  0.421634
5   2022484408                  4.530196
6   2026352035                  3.613338
7   2320127002                  2.735748
8   2347167796                  4.658637
9   2873212765                  3.941524
10  3372868164                  3.549558
11  3977333714                  7.256926
12  4020332650                  0.950297
13  4057192912                  1.944522
14  4319703577                  3.567218
15  4388161847                  3.495277
16  4445114986                  2.194018
17  4558609924                  3.779712
18  4702921684                  2.890550
19  5553957443                  4.591717
20  5577150313                  2.471827
21  6117666160                  3.116439
22  6290855005                  2.173222
23  6775888955                  1.181972
24  6962181067                  4.941800
25  7007744171                  4.451031
26  7086361926                  3.651784
27  8053475328                  5.011629
28  8253242879                  3.625368
29  8378563200                  2.536739
30  8583815059                  2.634858
31  8792009665                  0.944664
32  8877689391                  4.689714

### How heart rate correlates with activity level (light, moderate, very active)

In [43]:
df = pd.read_sql('''WITH heart_rate AS (
               SELECT
                    SUBSTR(Time, 1, 16) || '00' AS formatted_datetime,
                    AVG(Value) AS avg_heart_rate_in_one_min
               FROM heartrate_seconds_merged
               GROUP BY 1
               ),
               intensity AS (
                SELECT
                    SUBSTR(ActivityMinute, 1, 16) || '00' AS activity_min,
                    Intensity AS intensity
                FROM minuteIntensitiesNarrow_merged
               ),
               corr AS (
                SELECT
                    intensity.activity_min,
                    hr.avg_heart_rate_in_one_min,
                    intensity.intensity
                FROM heart_rate hr
                JOIN intensity
                ON hr.formatted_datetime = intensity.activity_min
               )
               SELECT * FROM corr
               ''', conn)

# Calculate Pearson correlation coefficient
correlation_coefficient = df['avg_heart_rate_in_one_min'].corr(df['intensity'], method='pearson')
correlation_coefficient

0.06929728074040427

#### The correlation of 0.069 indicates a weak positive connection between heart rate and activity intensity (light, moderate, very active). This modest link is influenced by differences in fitness levels, sample variations, and other factors. Understanding individual differences is essential when studying heart rate and physical activity.